### PR TITLE
feat(opentelemtry): Initial implementation of OpenTelemetry

### PR DESCRIPTION
**Pull request by phiz71**
---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `1.0.0`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/tracer/gravitee-tracer-opentelemetry/1.0.0/gravitee-tracer-opentelemetry-1.0.0.zip)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    <artifactId>gravitee-tracer-opentelemetry</artifactId>
+    <version>1.0.0</version>
+
+    <name>Gravitee.io - Tracing - OpenTelemetry OTLP</name>
+
+    <properties>
+        <gravitee-bom.version>8.1.0</gravitee-bom.version>
+        <gravitee-node-api.version>5.18.0</gravitee-node-api.version>
+        <opentelemetry-instrumentation-bom-alpha.version>1.33.3-alpha</opentelemetry-instrumentation-bom-alpha.version>
+        <opentelemetry-instrumentation-api.version>1.33.3</opentelemetry-instrumentation-api.version>
```
Comment (phiz71) [inline pom.xml:40]: Is it redundant ?

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            <artifactId>opentelemetry-semconv</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry.instrumentation</groupId>
+            <artifactId>opentelemetry-runtime-telemetry-java17</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-api-incubator</artifactId>
+            <version>1.39.0-alpha</version>
```
Comment (phiz71) [inline pom.xml]: Should be in `<properties/>`

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        <!-- Vert.x dependencies -->
+        <dependency>
+            <groupId>io.vertx</groupId>
+            <artifactId>vertx-core</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.vertx</groupId>
+            <artifactId>vertx-grpc-client</artifactId>
+            <version>${vertx.version}</version>
```
Comment (phiz71) [inline pom.xml]: Isn't already declared in gravitee-bom>vertx-dependencies ?

File: README.adoc [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+endif::[]
+
+== Description
+The `otel` tracer is used to collect tracing information from the gateway and send them to a compatible OpenTelemetry exporter (Jaeger, Datadog, Elastic APM, NewRelic, ...).
+
+== Installation
+
+You can build the tracer plugin from the source or you can download it from https://download.gravitee.io/#graviteeio-apim/plugins/tracers/graviteeio-tracer-opentelemetry/[here]
+
+Then, put the ZIP file in your gateway plugins folder. (https://docs.gravitee.io/apim/3.x/apim_installguide_gateway_configuration.html#configure_the_plugins_repository[More information])
```
Comment (remibaptistegio) [inline README.adoc]:
```suggestion
Then, put the ZIP file in your gateway plugins folder. (https://documentation.gravitee.io/apim/overview/plugins[More information])
```
Maybe link to the new doc site no ?

File: src/main/java/io/gravitee/tracer/opentelemetry/configuration/OpenTelemetryTracerConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+public class OpenTelemetryTracerConfiguration {
+
+    @Value("${services.tracing.otel.traces.enabled:true}")
+    private boolean tracesEnabled = true;
+
+    @Value("${services.tracing.otel.metrics.enabled:false}")
+    private boolean metricsEnabled = false;
+
+    @Value("${services.tracing.otel.ssl.enabled:false}")
+    private boolean sslEnabled = true;
```
Comment (remibaptistegio) [inline src/main/java/io/gravitee/tracer/opentelemetry/configuration/OpenTelemetryTracerConfiguration.java]: Should be false like your default value no ?

File: src/main/java/io/gravitee/tracer/opentelemetry/OpenTelemetryTracer.java [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+    private io.opentelemetry.api.trace.Tracer tracer;
+
+    private List<InstrumenterVertxTracer<?, ?>> instrumenterVertxTracers;
+
+    private RuntimeMetrics runtimeMetrics;
+
+    private OpenTelemetrySdk openTelemetrySdk;
+
+    private static final String EXPORTER_NAME = "otlp";
```
Comment (remibaptistegio) [inline src/main/java/io/gravitee/tracer/opentelemetry/OpenTelemetryTracer.java]: Not used no ?

File: src/main/java/io/gravitee/tracer/opentelemetry/exporter/ExporterFactory.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                        continue;
+                    }
+                    String[] parts = header.split("=", 2);
+                    String key = parts[0].trim();
+                    String value = parts[1].trim();
+                    headersMap.put(key, value);
+                }
+            }
+        }
+         */
```
Comment (remibaptistegio) [inline src/main/java/io/gravitee/tracer/opentelemetry/exporter/ExporterFactory.java]: To remove I guess ?

File: pom.xml (hunk 3 above)
Comment (brasseld) [reply pom.xml]: No, it's not @phiz71

Comment (phiz71) [reply pom.xml]: https://github.com/vert-x3/vertx-dependencies/blob/4.5.7/pom.xml#L911-L914 ?

File: src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseGrpcExporter.java [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+                        // TODO: should we only retry on a specific errors?
+
+                        client
+                            .request(server)
+                            .onSuccess(onSuccessHandler)
+                            .onFailure(
+                                new Handler<>() {
+                                    @Override
+                                    public void handle(Throwable event) {
+                                        failOnClientRequest(numItems, t, result);
```
Comment (remibaptistegio) [inline src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseGrpcExporter.java:145]: Is it wanted to log the first Throwable and not the one from the retry ?

File: src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseGrpcExporter.java [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+                                        envVar = "OTEL_METRICS_EXPORTER";
+                                        break;
+                                    case "log":
+                                        envVar = "OTEL_LOGS_EXPORTER";
+                                        break;
+                                    default:
+                                        throw new IllegalStateException(
+                                            "Unrecognized type, this is a programming bug in the OpenTelemetry SDK"
+                                        );
+                                }
```
Comment (remibaptistegio) [inline src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseGrpcExporter.java]:
```suggestion
                                String envVar =
                                    switch (type) {
                                        case "span" -> "OTEL_TRACES_EXPORTER";
                                        case "metric" -> "OTEL_METRICS_EXPORTER";
                                        case "log" -> "OTEL_LOGS_EXPORTER";
                                        default -> throw new IllegalStateException(
                                            "Unrecognized type, this is a programming bug in the OpenTelemetry SDK"
                                        );
                                    };
```
New syntax adds readability, wdyt ?

File: src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseHttpExporter.java [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+                marshaler,
+                1
+            );
+            initiateSend(client, requestURI, MAX_ATTEMPTS, clientRequestSuccessHandler, onError);
+        }
+
+        private static void initiateSend(
+            HttpClient client,
+            String requestURI,
+            int numberOfAttempts,
```
Comment (remibaptistegio) [inline src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseHttpExporter.java:140]: It seems that you are not using this parameter. You can remove it and just use the MAX_ATTEMPT with the attemptNumber from the ClientRequestSuccessHandler.

**Review by brasseld — APPROVED** [review]

File: src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseGrpcExporter.java (hunk 8 above)
Comment (brasseld) [reply src/main/java/io/gravitee/tracer/opentelemetry/exporter/BaseGrpcExporter.java:145]: yes !

**Review by remibaptistegio — APPROVED** [review]

**Pull request merged**